### PR TITLE
Better ranges for implicit `Inherit` error reporting.

### PR DESCRIPTION
**Pull request by edgarfgp**
## Description

Fixes # (issue, if applicable)

## Checklist

- [x] Test cases added
- [x] Release notes entry updated

### CODE REVIEW

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 1]
```diff
@@ -3335,21 +3334,23 @@ module EstablishTypeDefinitionCores =
 
                       | [(ty, m)] ->
                           let inheritRange =
-                                match inherits with
-                                | [] -> m
-                                | (synType, _, _) :: _ -> synType.Range
+                              match inherits with
+                              | [] -> m
+                              | (synType, _, _) :: _ -> synType.Range
```
Comment (T-Gro) [inline src/Compiler/Checking/CheckDeclarations.fs:3339]:
```suggestion
                              | (synType, _, _) :: _ -> synType.Range
```

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 2]
```diff
@@ -3335,21 +3334,23 @@ module EstablishTypeDefinitionCores =
...
                               Some g.obj_ty_noNulls // a "super" that is a variable type causes grief later
-                          else             
-                                       
+                          else                  
                               Some ty 
-                      | _ -> 
-                          error(Error(FSComp.SR.tcTypesCannotInheritFromMultipleConcreteTypes(), m))
+                      | _ ->
+                          inherits
+                          |> List.skip 1
```
Comment (T-Gro) [inline src/Compiler/Checking/CheckDeclarations.fs:3351]: Please lean on the compiler to make sure the list length will be valid, instead of the universal `_` clause together with potentially unsafe `List.skip` (it is not unsafe now, but leaning on the pattern match completeness will be more robust against future refactorings)

Like ` | _first :: superFluousInherits `

File: docs/release-notes/.FSharp.Compiler.Service/9.0.200.md [hunk 3]
```diff
@@ -22,5 +22,6 @@
 * Better ranges for CE `use` error reporting. ([PR #17811](https://github.com/dotnet/fsharp/pull/17811))
 * Better ranges for `inherit` error reporting. ([PR #17879](https://github.com/dotnet/fsharp/pull/17879))
 * Better ranges for `inherit` `struct` error reporting. ([PR #17886](https://github.com/dotnet/fsharp/pull/17886))
+* Better ranges for implicit `inherit` error reporting. ([PR #17893](https://github.com/dotnet/fsharp/pull/17893))
```
Comment (psfinaki) [inline docs/release-notes/.FSharp.Compiler.Service/9.0.200.md:25]: Sorry for picking on the words - what is _implicit_ about this type of inheritance? I think implicitly we only inherit Object, if nothing is specified 🤔

Comment (edgarfgp) [reply docs/release-notes/.FSharp.Compiler.Service/9.0.200.md:25]: It is how is named in the `AST` `SynMemberDefn.ImplicitInherit`. I guess we use `inherit objects` instead. ?

File: src/Compiler/Checking/CheckDeclarations.fs (hunk 2 above)
Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs]: Sure.

File: docs/release-notes/.FSharp.Compiler.Service/9.0.200.md (hunk 3 above)
Comment (edgarfgp) [reply docs/release-notes/.FSharp.Compiler.Service/9.0.200.md]: Updated.

**Review by psfinaki — APPROVED** [review]
Thanks Edgar!

File: tests/FSharp.Compiler.ComponentTests/ErrorMessages/ClassesTests.fs [hunk 4]
```diff
@@ -849,4 +849,39 @@ type C5 = class inherit System.MulticastDelegate override x.ToString() = ""  end
...
+
+type Class() =
+    inherit ClassA()
+    inherit ClassB()
+    inherit ClassC()
+        """
+        |> typecheck
+        |> shouldFail
+        |> withDiagnostics [
+            (Error 959, Line 10, Col 5, Line 10, Col 12, "Type definitions may only have one 'inherit' specification and it must be the first declaration")
```
[the remark this answers is not in the document]
Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/ErrorMessages/ClassesTests.fs:883]: @abonie 

> Do we actually want both these errors? I mean 959 and 932 for each additional inherit?

I guess arguably we do not need 959 as 932 will cover the cases for having multiple `inherit`.  Should we get ride of 959 ? 

Personally I think showing the specific ranges for inherit is more clear and guides the user toward removing rather than just using the type definition.

**Review by T-Gro — APPROVED** [review]

**Comment by edgarfgp** [comment]
Please wait for https://github.com/dotnet/fsharp/pull/17905/files to be merged first.

File: docs/release-notes/.FSharp.Compiler.Service/9.0.200.md
Comment (psfinaki) [inline docs/release-notes/.FSharp.Compiler.Service/9.0.200.md:1]: Marker to not merge this accidentally.

**Comment by auduchinok** [comment]
@edgarfgp Thanks for your contribution! I'm sorry to say this, but I disagree with this change. I'll try to explain it on this example:
```fsharp
type T() =
    inherit A()
    inherit B()
```

I expect that there's an error about multiple `inherit` members. That means that the actual source of the errors are the `inherit` members themselves, not the type names inside of them. Moving the error to the inherited type name may say to many people that there's no error with the inherit member itself, and there's something wrong with the inherited type usage.

**Comment by auduchinok** [comment]
Could you please also add tests for multiple base interfaces, if there're none?

```fsharp
type I =
    inherit IA
    inherit IB
```

**Comment by edgarfgp** [comment]
> @edgarfgp Thanks for your contribution! I'm sorry to say this, but I disagree with this change. I'll try to explain it on this example:
> 
> ```fsharp
> type T() =
>     inherit A()
>     inherit B()
> ```
> 
> I expect that there's an error about multiple `inherit` members. That means that the actual source of the errors are the `inherit` members themselves, not the type names inside of them. Moving the error to the inherited type name may say to many people that there's no error with the inherit member itself, and there's something wrong with the inherited type usage.

@auduchinok  No need to apologise for disagreeing. This is a collaborative effort and Im glad that you are reviewing my PRs :) 

So is the following what you would expect ? 

```fsharp 
[<Fact>]
    let ``Types cannot inherit from multiple concrete types.`` () =
        Fsx """
type ClassA() = class end

type ClassB() = class end

type ClassC() = class end

type Class() =
    inherit ClassA()
    inherit ClassB()
    inherit ClassC()
        """
        |> typecheck
        |> shouldFail
        |> withDiagnostics [
            (Error 959, Line 10, Col 5, Line 10, Col 12, "Type definitions may only have one 'inherit' specification and it must be the first declaration")
            (Error 959, Line 11, Col 5, Line 11, Col 12, "Type definitions may only have one 'inherit' specification and it must be the first declaration")
            (Error 932, Line 10, Col 13, Line 10, Col 19, "Types cannot inherit from multiple concrete types")
            (Error 932, Line 11, Col 13, Line 11, Col 19, "Types cannot inherit from multiple concrete types")
        ]
```

**Comment by auduchinok** [comment]
@edgarfgp More or less, yes. I'd probably prefer the `932` errors to be on the `Class` type name, not on the `inherit` members.

**Comment by edgarfgp** [comment]
> @edgarfgp More or less, yes. I'd probably prefer the `932` errors to be on the `Class` type name, not on the `inherit` members.

Humm. But using the type name and not the `inherit` range makes it look like multiple inherits are allowed(Unless you hover the type name). I would prefer to be use the accurate range.

**Comment by edgarfgp** [comment]
> @edgarfgp Thanks for your contribution! I'm sorry to say this, but I disagree with this change. I'll try to explain it on this example:
> 
> ```fsharp
> type T() =
>     inherit A()
>     inherit B()
> ```
> 
> I expect that there's an error about multiple `inherit` members. That means that the actual source of the errors are the `inherit` members themselves, not the type names inside of them. Moving the error to the inherited type name may say to many people that there's no error with the inherit member itself, and there's something wrong with the inherited type usage.

@auduchinok Updated following you suggestion. @abonie also [suggested](https://github.com/dotnet/fsharp/pull/17893#discussion_r1806748000) the same previously.

File: docs/release-notes/.FSharp.Compiler.Service/9.0.200.md
Comment (edgarfgp) [reply docs/release-notes/.FSharp.Compiler.Service/9.0.200.md:1]: This is now ready

File: tests/FSharp.Compiler.ComponentTests/ErrorMessages/ClassesTests.fs [hunk 5]
```diff
@@ -861,4 +861,83 @@ type C5 = class inherit System.MulticastDelegate override x.ToString() = ""  end
...
+
+type Class() =
+    inherit ClassA()
+    inherit ClassB()
+    inherit ClassC()
+        """
+        |> typecheck
+        |> shouldFail
+        |> withDiagnostics [
+            (Error 959, Line 10, Col 5, Line 10, Col 21, "Type definitions may only have one 'inherit' specification and it must be the first declaration")
```
Comment (auduchinok) [inline tests/FSharp.Compiler.ComponentTests/ErrorMessages/ClassesTests.fs:895]: What do you think about moving `959` to `Class` on line 8? From its text it seems to be about the type declaration having errors inside of it, not about some particular `inherit` member.

Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/ErrorMessages/ClassesTests.fs]: I think it makes sense as the error message mentions `Type definitions`.

Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/ErrorMessages/ClassesTests.fs]: Updated. See my last commit.

**Review by auduchinok — APPROVED** [review]

**Comment by auduchinok** [comment]
@edgarfgp Thanks!

File: tests/FSharp.Compiler.ComponentTests/ErrorMessages/ClassesTests.fs [hunk 6]
```diff
@@ -861,4 +861,83 @@ type C5 = class inherit System.MulticastDelegate override x.ToString() = ""  end
...
+type IB = interface end
+
+type I() =
+    inherit IA
+    inherit IB
+        """
+        |> typecheck
+        |> shouldFail
+        |> withDiagnostics [
+            (Error 961, Line 6, Col 5, Line 6, Col 12, "This 'inherit' declaration specifies the inherited type but no arguments. Consider supplying arguments, e.g. 'inherit BaseType(args)'.")
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/ErrorMessages/ClassesTests.fs:941]: This is a bit misleading here... but looks like it's been like this before, so just pointing out.

Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/ErrorMessages/ClassesTests.fs:941]: @psfinaki 

Yes it has been like this. What do you think we should do here ? . The main issue here is that you can not from an interface when using a class, only with a type.

Screenshot 2024-10-25 at 15 43 57

I think we should replace  `961` with  `error FS0946: Cannot inherit from interface type. Use interface ... with instead`

Comment (psfinaki) [reply tests/FSharp.Compiler.ComponentTests/ErrorMessages/ClassesTests.fs:941]: Yeah IMO the first error is just redundant. We show two errors now, I think we could leave just 946 here.
